barretenberg/cpp/src/barretenberg/plonk_honk_shared/library/grand_product_library.hpp

```diff
@@ -84,22 +87,34 @@ void compute_grand_product(typename Flavor::ProverPolynomials& full_polynomials,
     Polynomial numerator{ domain_size, domain_size };
     Polynomial denominator{ domain_size, domain_size };
 
+    auto check_is_active = [&](size_t idx) {
+        if (active_block_ranges.empty()) {
+            return true;
+        }
+        return std::any_of(active_block_ranges.begin(), active_block_ranges.end(), [idx](const auto& range) {
+            return idx >= range.first && idx < range.second;
+        });
+    };
+
     // Step (1)
     // Populate `numerator` and `denominator` with the algebra described by Relation
+    FF gamma_fourth = relation_parameters.gamma.pow(4);
     parallel_for(num_threads, [&](size_t thread_idx) {
-        typename Flavor::AllValues evaluations;
-        // TODO(https://github.com/AztecProtocol/barretenberg/issues/940): construction of evaluations is equivalent to
-        // calling get_row which creates full copies. avoid?
+        typename Flavor::AllValues row;
         const size_t start = idx_bounds[thread_idx].first;
         const size_t end = idx_bounds[thread_idx].second;
         for (size_t i = start; i < end; ++i) {
-            for (auto [eval, full_poly] : zip_view(evaluations.get_all(), full_polynomials.get_all())) {
-                eval = full_poly.size() > i ? full_poly[i] : 0;
+            if (check_is_active(i)) {
```

Comment:
I suppose by avoiding unnecessary calls to get_row you've cut down most of the additional cost but there's more we could potentially do right? IIUC with this update we still go on to compute z_perm everywhere even in the "constant" ranges which could be avoided. Maybe that's a negligible cost compared to what you've handled here. One thought: maybe its not this easy but could we simply update the `parallel_for` loop that appears multiple times in this method to operate only over the active ranges? We shouldn't really even need to compute any intermediate values in the inactive regions. We would just populate the constant regions with the appropriate constant value at the end. Anyway, I'll leave it to you to decide what to do if anything.

Reply:
yes, thanks for the comment, will have a go at it